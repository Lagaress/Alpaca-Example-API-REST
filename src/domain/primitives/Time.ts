import moment, { DurationInputArg1, DurationInputArg2 } from 'moment';

export default class Time {
  private static readonly MILLISECONDS_TO_SECONDS = 1000;

  private readonly milliseconds: number;

  private constructor(milliseconds: number) {
    this.milliseconds = milliseconds;
  }

  public valueOf(): number {
    return this.milliseconds;
  }

  public valueOfSeconds() {
    return Math.round(this.milliseconds / Time.MILLISECONDS_TO_SECONDS);
  }

  public toUnix(): number {
    return moment.utc(this.valueOf()).unix();
  }

  public toDate(): Date {
    return moment.utc(this.valueOf()).toDate();
  }

  public format(format?: string): string {
    return moment.utc(this.valueOf()).format(format);
  }

  public toString(): string {
    return this.format();
  }

  public isAfter(time: Time): boolean {
    return moment.utc(this.valueOf()).isAfter(time.valueOf());
  }

  public isBefore(time: Time): boolean {
    return moment.utc(this.valueOf()).isBefore(time.valueOf());
  }

  public get isFuture(): boolean {
    return moment.utc().isBefore(this.valueOf());
  }

  public get isPast(): boolean {
    return !this.isFuture;
  }

  public get isToday(): boolean {
    return moment.utc().isSame(moment.utc(this.valueOf()), 'day');
  }

  public add(amount: DurationInputArg1, unit: DurationInputArg2): Time {
    return Time.from(moment.utc(this.valueOf()).add(amount, unit).valueOf());
  }

  public subtract(amount: DurationInputArg1, unit: DurationInputArg2): Time {
    return Time.from(moment.utc(this.valueOf()).subtract(amount, unit).valueOf());
  }

  public static from(milliseconds: number): Time {
    return new Time(milliseconds);
  }

  public static fromString(timestamp: string): Time {
    return Time.from(moment.utc(timestamp).valueOf());
  }

  public static timestamp(milliseconds?: number): Time {
    return milliseconds ? Time.from(milliseconds) : Time.now();
  }

  public static optional(milliseconds?: number): Time | null {
    return milliseconds ? Time.from(milliseconds) : null;
  }

  public static optionalFromString(timestamp?: string): Time | null {
    return timestamp ? Time.fromString(timestamp) : null;
  }

  public static fromSeconds(seconds: number): Time {
    return Time.from(seconds * Time.MILLISECONDS_TO_SECONDS);
  }

  public static now(): Time {
    return Time.from(moment.utc().valueOf());
  }
}
