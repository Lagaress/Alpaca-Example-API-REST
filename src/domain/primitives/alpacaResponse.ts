type AlpacaResponseParams = {
  code: string;
  statusCode: number;
  message: string;
}

export default class AlpacaResponse {
  public code: string;
  public statusCode: number;
  public message: string;

  constructor(response: AlpacaResponseParams) {
    this.code = response.code;
    this.statusCode = response.statusCode;
    this.message = response.message;
  }
}

