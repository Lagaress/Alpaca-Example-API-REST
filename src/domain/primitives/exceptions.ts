import statusCodes from 'http-status-codes';
import AlpacaResponse from './alpacaResponse';

export const UserNotFound = new AlpacaResponse({ code: 'US40401', statusCode: statusCodes.BAD_REQUEST, message: 'The user does not exist' });

export const UserAlreadyExists = new AlpacaResponse({ code: 'US40901', statusCode: statusCodes.CONFLICT, message: 'User already exists' });

export const InternalError = new AlpacaResponse({ code: 'ER50001', statusCode: statusCodes.INTERNAL_SERVER_ERROR, message: 'Internal server error' });
