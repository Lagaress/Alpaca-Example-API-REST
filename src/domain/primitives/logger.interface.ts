interface LogFunction {
  (message: string): void;
  (object: Record<string, unknown>, message?: string): void;
}

interface LoggerInterface {
  trace: LogFunction;
  debug: LogFunction;
  info: LogFunction;
  warn: LogFunction;
  error: LogFunction;
  fatal: LogFunction;
  child(object: Record<string, unknown>): LoggerInterface;
}

export default LoggerInterface;
