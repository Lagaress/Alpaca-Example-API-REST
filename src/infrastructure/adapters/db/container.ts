import { AwilixContainer, asValue, createContainer, asFunction } from 'awilix';
import mysql2 from 'mysql2';
import knex from 'knex';

import dbAdapter from './db.adapter';

const container: AwilixContainer = createContainer();

container.register({
  mysql2: asValue(mysql2),
  knex: asValue(knex),
  dbAdapter: asFunction(dbAdapter).singleton(),
});

export default container;

export type Dependencies = {
  // Libraries
  mysql2: typeof mysql2;
  knex: typeof knex;

  // adapter
  dbAdapter: ReturnType<typeof dbAdapter>;
};
