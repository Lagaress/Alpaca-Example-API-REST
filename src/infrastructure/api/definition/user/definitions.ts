export default {
  user: {
    type: 'object',
    properties: {
      id: {
        description: 'User ID',
        type: 'number',
        example: 53,
      },
      email: {
        description: 'User email',
        type: 'string',
        example: 'john.doe@example.com',
      },
      nickname: {
        description: 'User nickname',
        type: 'string',
        example: 'John Doe',
      },
      createdAt: {
        description: 'The date when the user was created',
        type: 'string',
        example: '2020-01-13 09:45:01',
      },
      updatedAt: {
        description: 'The date when the user was updated',
        type: 'string',
        example: '2020-01-13 09:45:01',
      },
    },
  },

  userResponse: {
    allOf: [
      {
        $ref: '#/definitions/apiResult',
      },
      {
        type: 'object',
        properties: {
          data: {
            type: 'object',
            $ref: '#/definitions/user',
          },
        },
      },
    ],
  },

  userPaginatedResponse: {
    allOf: [
      {
        $ref: '#/definitions/apiResult',
      },
      {
        type: 'object',
        properties: {
          data: {
            type: 'object',
            properties: {
              data: {
                type: 'array',
                items: {
                  $ref: '#/definitions/user',
                },
              },
            },
          },
          extra: {
            type: 'object',
            $ref: '#/definitions/paginationInformation',
          },
        },
      },
    ],
  },
};
