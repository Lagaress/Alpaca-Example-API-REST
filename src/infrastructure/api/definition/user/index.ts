export default {
  '/users': {
    post: {
      operationId: 'create',
      'x-swagger-router-controller': 'user.entrypoint',
      tags: [
        'users',
      ],
      description: 'Creates a new user',
      parameters: [
        {
          name: 'data',
          in: 'body',
          required: true,
          schema: {
            type: 'object',
            additionalProperties: false,
            required: [ 'email', 'nickname' ],
            properties: {
              email: {
                description: 'User email',
                type: 'string',
                example: 'johndoe@example.com',
              },
              nickname: {
                description: 'User nickname',
                type: 'string',
                example: 'johndoe',
              },
            },
          },
        },
      ],
      responses: {
        201: {
          description: 'User created successfully',
          schema: {
            $ref: '#/definitions/userResponse',
          },
        },
        500: {
          description: 'Internal server error',
          schema: {
            $ref: '#/definitions/apiResult',
          },
        },
      },
    },
    get: {
      operationId: 'getAll',
      'x-swagger-router-controller': 'user.entrypoint',
      tags: [
        'users',
      ],
      description: 'Retrieves all users',
      parameters: [
        {
          name: 'nickname',
          in: 'query',
          type: 'string',
          description: 'User nickname',
        },
        {
          name: 'email',
          in: 'query',
          type: 'string',
          description: 'User email',
        },
        { $ref: '#/parameters/limit' },
        { $ref: '#/parameters/offset' },
      ],
      responses: {
        200: {
          description: 'Users retrieved successfully',
          schema: {
            $ref: '#/definitions/userPaginatedResponse',
          },
        },
        500: {
          description: 'Internal server error',
          schema: {
            $ref: '#/definitions/apiResult',
          },
        },
      },
    },
  },

  '/users/{userId}': {
    parameters: [
      {
        name: 'userId',
        in: 'path',
        required: true,
        type: 'number',
        description: 'User id',
      },
    ],
    get: {
      operationId: 'get',
      'x-swagger-router-controller': 'user.entrypoint',
      tags: [
        'users',
      ],
      description: 'Retrieves a user by id',
      responses: {
        200: {
          description: 'User retrieved successfully',
          schema: {
            $ref: '#/definitions/userResponse',
          },
        },
        404: {
          description: 'User not found',
          schema: {
            $ref: '#/definitions/apiResult',
          },
        },
        500: {
          description: 'Internal server error',
          schema: {
            $ref: '#/definitions/apiResult',
          },
        },
      },
    },
    patch: {
      operationId: 'update',
      'x-swagger-router-controller': 'user.entrypoint',
      tags: [
        'users',
      ],
      description: 'Updates a user by id',
      parameters: [
        {
          name: 'data',
          in: 'body',
          required: true,
          schema: {
            type: 'object',
            additionalProperties: false,
            properties: {
              email: {
                description: 'User email',
                type: 'string',
                example: 'johndoe@example.com',
              },
              nickname: {
                description: 'User nickname',
                type: 'string',
                example: 'johndoe',
              },
            },
          },
        },
      ],
      responses: {
        202: {
          description: 'User updated successfully',
          schema: {
            $ref: '#/definitions/userResponse',
          },
        },
        404: {
          description: 'User not found',
          schema: {
            $ref: '#/definitions/apiResult',
          },
        },
        500: {
          description: 'Internal server error',
          schema: {
            $ref: '#/definitions/apiResult',
          },
        },
      },
    },
    delete: {
      operationId: 'delete',
      'x-swagger-router-controller': 'user.entrypoint',
      tags: [
        'users',
      ],
      description: 'Deletes a user by id',
      responses: {
        204: {
          description: 'User deleted successfully',
        },
        500: {
          description: 'Internal server error',
          schema: {
            $ref: '#/definitions/apiResult',
          },
        },
      },
    },
  },
};

