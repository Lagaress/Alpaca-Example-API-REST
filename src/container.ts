import { AwilixContainer, Lifetime, asValue, createContainer } from 'awilix';

import axios from 'axios';
import cors from 'cors';
import express from 'express';
import helmet from 'helmet';
import knex from 'knex';
import mysql2 from 'mysql2';
import pino from 'pino';
import swaggerTools from 'swagger-tools';

import LoggerInterface from './domain/primitives/logger.interface';

import config from './config';
import { DbAdapterInterface } from './infrastructure/adapters/db/db.adapter';
import dbAdapter from './infrastructure/adapters/db/index';
import serverAdapter from './infrastructure/adapters/server/server.adapter';
import healthController from './infrastructure/controllers/health.controller';
import userController from './infrastructure/controllers/user.controller';
import userService from './services/user/user.service';
import UserRepositoryInterface from './usecases/abstractions/userRepository.interface';
import userUsecases from './usecases/user/user.usecases';

const container: AwilixContainer = createContainer();

container.loadModules(
  [
    `${__dirname}/**/*.usecases.{ts,js}`,
    `${__dirname}/**/*.adapter.{ts,js}`,
    `${__dirname}/**/*.facade.{ts,js}`,
    `${__dirname}/**/*.factory.{ts,js}`,
    `${__dirname}/**/logger.{ts,js}`,
    `${__dirname}/**/*.controller.{ts,js}`,
    `${__dirname}/**/*.middleware.{ts,js}`,
    `${__dirname}/**/*.routes.{ts,js}`,
    `${__dirname}/**/*.repository.{ts,js}`,
    `${__dirname}/**/*.service.{ts,js}`,
    `${__dirname}/**/*.handler.{ts,js}`,
    `${__dirname}/**/*.proxy.{ts,js}`,
  ],
  {
    formatName: 'camelCase',
    resolverOptions: {
      lifetime: Lifetime.SINGLETON,
    },
  }
);

container.register({
  express: asValue(express),
  swaggerTools: asValue(swaggerTools),
  pino: asValue(pino),
  axios: asValue(axios),
  mysql2: asValue(mysql2),
  knex: asValue(knex),
  cors: asValue(cors),
  helmet: asValue(helmet),
  dbAdapter: asValue(dbAdapter(config.DB)),
});

export default container;

export type Dependencies = {
  // Libraries
  express: typeof express;
  swaggerTools: typeof swaggerTools;
  pino: typeof pino;
  axios: typeof axios;
  mysql2: typeof mysql2;
  knex: typeof knex;
  cors: typeof cors;
  helmet: typeof helmet;

  // Definitions
  logger: LoggerInterface;

  // Adapters
  serverAdapter: ReturnType<typeof serverAdapter>;
  dbAdapter: DbAdapterInterface;

  // Controllers
  userController: ReturnType<typeof userController>;
  healthController: ReturnType<typeof healthController>;

  // Repositories
  userRepository: UserRepositoryInterface;

  // Usecases
  userUsecases: ReturnType<typeof userUsecases>;

  // Services
  userService: ReturnType<typeof userService>;
};
