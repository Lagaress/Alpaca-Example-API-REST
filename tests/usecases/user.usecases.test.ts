import { UserAlreadyExists, UserNotFound } from "../../src/domain/primitives/exceptions";
import { UserFilter } from "../../src/usecases/abstractions/userRepository.interface";
import { EntityPagination } from "../../src/usecases/types/queryParams.types";
import { UserCreationPayload, UserUpdationPayload } from "../../src/usecases/user/user.usecases";
import { UserBuilder } from "../builders/user.builder";
import userRepositoryStub from "../stubs/user.repository.stub";
import container from "../test.container";

const { get, getAll, create, update, delete: deleteUser } = container.resolve('userUsecases');

describe('User Usecases', () => {
  const user = new UserBuilder().build();

  describe('Create', () => {
    const payload: UserCreationPayload = {
      email: 'test@test.com',
      nickname: 'test',
      password: 'test',
    };
    const newUser = new UserBuilder()
      .with('email', 'test@test.com')
      .with('nickname', 'test')
      .with('password', 'test')
      .build();

    beforeEach(async () => {
      userRepositoryStub.create.mockReset().mockResolvedValueOnce(newUser);
      userRepositoryStub.getAll.mockReset().mockResolvedValueOnce([]);
    });

    it('should retrieve the existing user from the repository', async () => {
      await create(payload);
      expect(userRepositoryStub.create).toHaveBeenCalledTimes(1);
      expect(userRepositoryStub.getAll).toHaveBeenCalledWith({ email: payload.email });
    });

    it('should fail if the user already exists', async () => {
      userRepositoryStub.getAll.mockResolvedValueOnce([user]);
      try {
        await create(payload);
      } catch (error) {
        expect(error).toBe(UserAlreadyExists);
      }
    });

    it('should create the user if it does not exist', async () => {
      await create(payload);
      expect(userRepositoryStub.create).toHaveBeenCalledTimes(1);
      expect(userRepositoryStub.create).toHaveBeenCalledWith(expect.objectContaining({
        email: payload.email,
        nickname: payload.nickname,
        password: payload.password,
      }));
    });

    it('should model it properly', async () => {
      const user = await create(payload);
      expect(user).toMatchSnapshot();
    });
  });

  describe('Get', () => {
    const userId = user.id;

    beforeEach(() => {
      userRepositoryStub.get.mockReset().mockResolvedValueOnce(user);
    });

    it('should retrieve the user from the repository', async () => {
      await get(userId);
      expect(userRepositoryStub.get).toHaveBeenCalledTimes(1);
      expect(userRepositoryStub.get).toHaveBeenCalledWith(userId);
    });

    it('should fail if the user does not exist', async () => {
      userRepositoryStub.get.mockReset().mockResolvedValueOnce(null);
      try {
        await get(userId);
      } catch (error) {
        expect(error).toBe(UserNotFound);
      }
    });

    it('should model it properly', async () => {
      const user = await get(userId);
      expect(user).toMatchSnapshot();
    });
  });

  describe('Get All', () => {
    const filter: UserFilter = {
      nickname: 'test',
    };
    const paginationParams: EntityPagination = {
      limit: 10,
      offset: 0,
    };

    const users = [user];

    beforeEach(() => {
      userRepositoryStub.getAll.mockReset().mockResolvedValueOnce(users);
      userRepositoryStub.count.mockReset().mockResolvedValueOnce(users.length);
    });
    
    it('should retrieve all users from the repository', async () => {
      await getAll(filter, paginationParams);
      expect(userRepositoryStub.getAll).toHaveBeenCalledTimes(1);
      expect(userRepositoryStub.getAll).toHaveBeenCalledWith(filter, paginationParams);
    });

    it('should count the total number of users', async () => {
      await getAll(filter, paginationParams);
      expect(userRepositoryStub.count).toHaveBeenCalledTimes(1);
      expect(userRepositoryStub.count).toHaveBeenCalledWith(filter);
    });

    it('should model it properly', async () => {
      const users = await getAll(filter, paginationParams);
      expect(users).toMatchSnapshot();
    });
  });

  describe('Update', () => {
    const userId = user.id;
    const payload: UserUpdationPayload = {
      email: 'test@test.com',
      nickname: 'test',
    };

    beforeEach(() => {
      userRepositoryStub.get.mockReset().mockResolvedValueOnce(user);
      userRepositoryStub.update.mockReset().mockResolvedValueOnce();
    });

    it('should retrieve the user from the repository', async () => {
      await update(userId, payload);
      expect(userRepositoryStub.get).toHaveBeenCalledTimes(1);
      expect(userRepositoryStub.get).toHaveBeenCalledWith(userId);
    });

    it('should fail if the user does not exist', async () => {
      userRepositoryStub.get.mockReset().mockResolvedValueOnce(null);
      try {
        await update(userId, payload);
      } catch (error) {
        expect(error).toBe(UserNotFound);
      }
    });

    it('should update the user', async () => {
      await update(userId, payload);
      expect(userRepositoryStub.update).toHaveBeenCalledTimes(1);
      expect(userRepositoryStub.update).toHaveBeenCalledWith(expect.objectContaining({
        id: userId,
        email: payload.email,
        nickname: payload.nickname,
      }));
    });
  });

  describe('Delete', () => {
    const userId = user.id;

    beforeEach(() => {
      userRepositoryStub.get.mockReset().mockResolvedValueOnce(user);
      userRepositoryStub.delete.mockReset();
    });

    it('should retrieve the user from the repository', async () => {
      await deleteUser(userId);
      expect(userRepositoryStub.get).toHaveBeenCalledTimes(1);
      expect(userRepositoryStub.get).toHaveBeenCalledWith(userId);
    });

    it('should delete the user', async () => {
      await deleteUser(userId);
      expect(userRepositoryStub.delete).toHaveBeenCalledTimes(1);
      expect(userRepositoryStub.delete).toHaveBeenCalledWith(userId);
    });

    it('should do nothing if the user does not exist', async () => {
      userRepositoryStub.get.mockReset().mockResolvedValueOnce(null);
      await deleteUser(userId);
      expect(userRepositoryStub.delete).not.toHaveBeenCalled();
    });
  }); 
});