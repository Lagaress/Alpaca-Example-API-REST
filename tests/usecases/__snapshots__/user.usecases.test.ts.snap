// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Usecases Create should model it properly 1`] = `
Object {
  "createdAt": "1970-03-05T05:02:42Z",
  "email": "test@test.com",
  "id": 12412,
  "nickname": "test",
  "updatedAt": "1970-03-05T05:02:42Z",
}
`;

exports[`User Usecases Get All should model it properly 1`] = `
Object {
  "total": 1,
  "users": Array [
    Object {
      "createdAt": "1970-03-05T05:02:42Z",
      "email": "john.doe@example.com",
      "id": 12412,
      "nickname": "John Doe",
      "updatedAt": "1970-03-05T05:02:42Z",
    },
  ],
}
`;

exports[`User Usecases Get should model it properly 1`] = `
Object {
  "createdAt": "1970-03-05T05:02:42Z",
  "email": "john.doe@example.com",
  "id": 12412,
  "nickname": "John Doe",
  "updatedAt": "1970-03-05T05:02:42Z",
}
`;
